### PR TITLE
Templatize incomplete content

### LINKED ISSUE #14: `_onTap` event listener doesn't handle correctly Shadow DOM path
In the onTap event `closest` function is used to make that the clicked button's closest data-nav is the current element. In 2.x with Shadow DOM `closest` function stops at shadowRoot and the event is not handled.

### LINKED ISSUE #15: cosmoz-data-nav should provide incomplete content
cosmoz-data-nav should provide incomplete content from a default template with the option to override it with another template.

### PR DESCRIPTION
**Pull request by megheaiulian**
Templatize incomplete content

fixes #15 
fixes #14

### CODE REVIEW

File: cosmoz-data-nav.js [hunk 1]
```diff
@@ -180,39 +179,98 @@
...
 
-		_onNodesChange() {
-			if (this._userTemplate) {
+		_onTemplatesChange(change) {
+			if (this._elementTemplate) {
 				return;
 			}
-			const template = this.queryEffectiveChildren('template');
+			const templates = change.addedNodes.filter(n => n.nodeType === Node.ELEMENT_NODE && n.tagName === 'TEMPLATE'),
+				elementTemplate = templates.find(n => n.matches(':not(incomplete)')),
```
Comment (nomego) [inline cosmoz-data-nav.js]: shouldn't the query be `':not([incomplete])'` ?

File: cosmoz-data-nav.js [hunk 2]
```diff
@@ -180,39 +179,98 @@
...
-			this._ensureTemplatized();
+			this._elementTemplate = elementTemplate;
+			this._incompleteTemplate = incompleteTemplate;
+			let baseProps = {
+				prevDisabled: true,
+				nextDisabled: true,
+				[this.indexAs]: true
+			};
+			this._elementCtor = Cosmoz.Templatize.templatize(this._elementTemplate, this, {
+				instanceProps: Object.assign({[this.indexAs]: true}, baseProps),
```
Comment (nomego) [inline cosmoz-data-nav.js]: indexAs already part of baseProps ?

File: cosmoz-data-nav.js [hunk 3]
```diff
@@ -180,39 +179,98 @@
...
+				forwardParentProp: this._forwardHostProp,
+				forwardParentPath: this._forwardParentPath,
+				forwardHostProp: this._forwardHostProp,
+			});
 
 			Array(this._elementsBuffer).fill(null).forEach(this._createElement, this);
+		},
+
+		get _allInstances() {
+			return this._elements.reduce((p, n) => p.concat([n.__instance, n.__incomplete]), []).filter(i => i != null);
```
Comment (nomego) [inline cosmoz-data-nav.js]: Do multi-line, same below

File: cosmoz-data-nav.js [hunk 4]
```diff
@@ -180,39 +179,98 @@
...
+
+		_forwardHostProp(prop, value) {
+			const instances = this._allInstances;
+			if (!instances || !instances.length) {
+				return;
+			}
+			instances.forEach(inst => IS_V2 ? inst.forwardHostProp(prop, value) : inst[prop] = value);
+		},
+
+		_notifyInstanceProp: function (inst, prop, value) {
```
Comment (nomego) [inline cosmoz-data-nav.js]: Use shorthand function decl

**Review by nomego — APPROVED** [review]

File: cosmoz-data-nav.js [hunk 5]
```diff
@@ -515,13 +533,13 @@
 				return;
 			}
 			const path = Polymer.dom(event).path,
-				attr = this.selectAttribute;
+				attr = this.selectAttribute,
+				selectEl = path.find(e => e && e.hasAttribute && e.hasAttribute(attr));
 
-			let select = path.find(e => e && e.hasAttribute && e.hasAttribute(attr));
-			if (!select || select.closest(this.is) !== this) {
+			if (!selectEl || path.slice(path.indexOf(selectEl)).find(e => e.is && e.is === this.is) !== this) {
```
Comment (nomego) [inline cosmoz-data-nav.js]: Please extract `path.slice(path.indexOf(selectEl))` and the .find() to two different const's to describe them

File: cosmoz-data-nav.js [hunk 6]
```diff
@@ -3,7 +3,17 @@
 
 (function () {
 	'use strict';
-	const IS_V2 = Polymer.flush != null;
+	const _async = window.requestIdleCallback || window.requestAnimationFrame || Polymer.Base.async,
+		_asyncPeriod = function (cb, minimum = 5) {
+			return _async(function (deadline) {
+				if (deadline && 'IdleDeadline' in window && deadline instanceof window.IdleDeadline && deadline.timeRemaining() < minimum) {
```
Comment (nomego) [inline cosmoz-data-nav.js]: We don't need the `function` keyword anymore :)
Also this if statement is way to long. Can you refactor somewhat for readability? At least a line break.

**Review by nomego — CHANGES_REQUESTED** [review]

File: cosmoz-data-nav.js (hunk 5 above)
Comment (plequang) [reply cosmoz-data-nav.js]: It should even be possible to replace the `find` + `indexOf` + `find` (again) combination with a single... `for` loop.
Just kidding 😂.

**Review by plequang — APPROVED** [review]
Looks good.
Waiting for latest changes requested by @nomego to merge.

**Review by nomego — APPROVED** [review]

**Pull request merged**